pkg/storaged/lvol-tabs.jsx

```diff
@@ -49,57 +48,145 @@ function lvol_rename(lvol) {
     });
 }
 
-function lvol_resize(client, lvol) {
+function lvol_and_fsys_resize(client, lvol, size, offline) {
+    var block, crypto, fsys;
+    var crypto_overhead;
+    var vdo;
+
+    block = client.lvols_block[lvol.path];
+    if (!block)
+        return lvol.Resize(size, { });
+
+    crypto = client.blocks_crypto[block.path];
+    if (crypto) {
+        var cleartext = client.blocks_cleartext[block.path];
+        if (!cleartext)
+            return;
+        fsys = client.blocks_fsys[cleartext.path];
+        vdo = client.vdo_overlay.find_by_backing_block(cleartext);
+        crypto_overhead = block.Size - cleartext.Size;
+    } else {
+        fsys = client.blocks_fsys[block.path];
+        vdo = client.vdo_overlay.find_by_backing_block(block);
+        crypto_overhead = 0;
+    }
+
+    function fsys_resize() {
+        if (fsys) {
+            // When doing an offline resize, we need to first repair the filesystem.
+            if (offline) {
+                return fsys.Repair({ }).then(function () { return fsys.Resize(size - crypto_overhead, { }); });
+            } else {
+                return fsys.Resize(size - crypto_overhead, { });
+            }
+        } else if (vdo) {
+            return vdo.grow_physical();
```

Comment:
What happens on shrink? If the function assumes that it's not ever called for shrinking VDO volumes, it should assert that.

Reply:
It is not called for shrinkin because `can_shrink` is false for VDO.  I'll add the assertion. What should happen in case the assertion fails?  Nothing, exception?